src/qtt/instrument_drivers/virtualAwg/awgs/ZurichInstrumentsHDAWG8.py

```diff
@@ -97,21 +107,47 @@ def update_gain(self, gain):
         The range is twice the gain under the assumption that the load on the output channels is 50 Ohm. For a high
         impedance load the gain equals the range.
         """
-        _ = [self.__awg.set('sigouts_{}_range'.format(ch), 2 * gain) for ch in self._channel_numbers]
+        _ = [self.__awg.set(f'sigouts_{ch}_range', 2 * gain) for ch in self._channel_numbers]
 
     def retrieve_gain(self):
-        gains = [self.__awg.get('sigouts_{}_range'.format(ch)) / 2 for ch in self._channel_numbers]
+        gains = [self.__awg.get(f'sigouts_{ch}_range') / 2 for ch in self._channel_numbers]
         if not all(g == gains[0] for g in gains):
             raise ValueError(f'Not all channel gains {gains} are equal. Please reset!')
         return gains[0]
 
+    @staticmethod
+    def waveform_to_wave(awg_driver, wave_name: str, waveform: np.ndarray) -> None:
+        """
+        Write waveforms to a .wave file in the modules data directory so that it
+        can be referenced and used in a sequence program.
+
+        Args:
+            wave_name: Name of the wave file, is used by a sequence program.
+            waveforms: Waveforms that is to be written to a .wave file.
+        """
+        data_dir = awg_driver.awg_module.getString('awgModule/directory')
+        wave_dir = os.path.join(data_dir, "awg", "waves")
+        if not os.path.isdir(wave_dir):
+            raise Exception(
+                "AWG module wave directory {} does not exist or is not a "
```

Comment:
f-string changes have been done by the autoformatter. Not sure why this one has not been changed. (perhaps because it is in an exception, the non f-string versions are more efficiency for logging purposes sometimes). I wil change it in the PR